app/services/service_limits.rb

```diff
@@ -0,0 +1,125 @@
+# ServiceLimits
+#
+# This class is the guard check for the HighlightsController CRUD actions
+class ServiceLimits
+  MAX_HIGHLIGHTS_PER_SOURCE_PER_USER = 300
+  MAX_HIGHLIGHTS_PER_USER            = 20_000
+  MAX_ANNOTATION_CHARS_PER_USER      = 300_000
+  MAX_CHARS_PER_ANNOTATION           = 1_000
+
+  class ServiceLimitsError < StandardError; end
+
+  class ExceededMaxHighlightsPerUser < ServiceLimitsError
+    def message
+      "Exceeded max highlights per user of #{MAX_HIGHLIGHTS_PER_USER}"
+    end
+  end
+
+  class ExceededMaxHighlightsPerUserPerSource < ServiceLimitsError
+    def message
+      "Exceeded max highlights per user per source of #{MAX_HIGHLIGHTS_PER_SOURCE_PER_USER}"
+    end
+  end
+
+  class ExceededMaxAnnotationCharsPerUser < ServiceLimitsError
+    def message
+      "Exceeded max annotation chars per user of #{MAX_ANNOTATION_CHARS_PER_USER}"
+    end
+  end
+
+  class ExceededMaxAnnotationChars < ServiceLimitsError
+    def message
+      "Exceeded max chars per annotation of #{MAX_CHARS_PER_ANNOTATION}"
+    end
+  end
+
+  def initialize(user_id:)
+    @user_id = user_id
+  end
+
+  def with_create_protection
+    ActiveRecord::Base.transaction do
+      yield(user).tap do |model|
+        raise ArgumentError, 'Block did not yield an active record model' unless model.is_a?(Highlight)
+        raise ActiveRecord::RecordInvalid.new(model) if model.invalid?
+
+        track_and_validate_max_highlights(model)
+        track_and_validate_annotation_chars(model) if model.annotation.present?
+      end
+    end
+  end
+
+  def with_update_protection
+    ActiveRecord::Base.transaction do
+      yield.tap do |model|
+        raise ArgumentError, 'Block did not yield an active record model' unless model.is_a?(Highlight)
+        raise ActiveRecord::RecordInvalid.new(model) if model.invalid?
+
+        prev_annotation_length = model.annotation_before_last_save.length
+        reset_max_for_annotations(user: user, by: -prev_annotation_length) if prev_annotation_length.present?
```

Comment:
Can we move these two lines into `track_and_validate_annotation_chars`?  I think we should be able to.